src/pip/_internal/metadata/pep723.py

```diff
@@ -0,0 +1,35 @@
+import re
+from typing import Any
+
+from pip._vendor import tomli as tomllib
+
+REGEX = r"(?m)^# /// (?P<type>[a-zA-Z0-9-]+)$\s(?P<content>(^#(| .*)$\s)+)^# ///$"
+
+
+class PEP723Exception(ValueError):
+    """Raised to indicate a problem when parsing PEP 723 metadata from a script"""
+
+    def __init__(self, msg: str) -> None:
+        self.msg = msg
+
+
+def pep723_metadata(scriptfile: str) -> dict[str, Any]:
+    with open(scriptfile) as f:
+        script = f.read()
+
+    name = "script"
+    matches = list(
+        filter(lambda m: m.group("type") == name, re.finditer(REGEX, script))
+    )
+    if len(matches) > 1:
+        raise PEP723Exception(f"Multiple {name} blocks found")
+    elif len(matches) == 1:
+        content = "".join(
+            line[2:] if line.startswith("# ") else line[1:]
+            for line in matches[0].group("content").splitlines(keepends=True)
+        )
+        return tomllib.loads(content)
```

Comment:
This can also raise an exception if the TOML is invalid.

Reply:
Good point, resolved in [f42ae07](https://github.com/pypa/pip/pull/13052/commits/f42ae0726a40afb49104da1624ac288d754da28e) and tested in [f459c7a](https://github.com/pypa/pip/pull/13052/commits/f459c7a4324cede34d96f8016ba2c0284430fb07) along with the multiple-blocks failure mode.